src/axom/quest/DistributedClosestPoint.hpp

```diff
@@ -442,186 +522,238 @@ class DistributedClosestPointImpl
     {
+      xferNodes[m_rank] = std::make_shared<conduit::Node>();
+      conduit::Node& xferNode = *xferNodes[m_rank];
+
       // clang-format off
-      auto& coords = mesh_node[fmt::format("coordsets/{}/values", coordset)];
+      auto& coords = queryMesh.fetch_existing(fmt::format("coordsets/{}/values", coordset));
       const int dim = internal::extractDimension(coords);
-      const int npts = internal::extractSize(coords);
-
-      xfer_node["npts"] = npts;
-      xfer_node["dim"] = dim;
-      xfer_node["src_rank"] = m_rank;
-      xfer_node["coords"].set_external(internal::getPointer<double>(coords["x"]), dim * npts);
-      xfer_node["cp_index"].set_external(internal::getPointer<axom::IndexType>(mesh_node["fields/cp_index/values"]), npts);
-      xfer_node["cp_rank"].set_external(internal::getPointer<axom::IndexType>(mesh_node["fields/cp_rank/values"]), npts);
-      xfer_node["closest_point"].set_external(internal::getPointer<double>(mesh_node["fields/closest_point/values/x"]), dim * npts);
-
-      if(mesh_node.has_path("fields/min_distance"))
+      const int qPtCount = internal::extractSize(coords);
+
+      xferNode["qPtCount"] = qPtCount;
+      xferNode["dim"] = dim;
+      xferNode["homeRank"] = m_rank;
+      xferNode["is_first"] = true;
+      xferNode["coords"].set_external(internal::getPointer<double>(coords["x"]), dim * qPtCount);
+      xferNode["cp_index"].set_external(internal::getPointer<axom::IndexType>(queryMesh.fetch_existing("fields/cp_index/values")), qPtCount);
+      xferNode["cp_rank"].set_external(internal::getPointer<axom::IndexType>(queryMesh.fetch_existing("fields/cp_rank/values")), qPtCount);
+      xferNode["closest_point"].set_external(internal::getPointer<double>(queryMesh.fetch_existing("fields/closest_point/values/x")), dim * qPtCount);
+      put_to_conduit_node(queryPartitionBb, xferNode["aabb"]);
+
+      if(queryMesh.has_path("fields/min_distance"))
       {
-        xfer_node["debug/min_distance"].set_external(internal::getPointer<double>(mesh_node["fields/min_distance/values"]), npts);
+        xferNode["debug/min_distance"].set_external(internal::getPointer<double>(queryMesh["fields/min_distance/values"]), qPtCount);
       }
       // clang-format on
-    }
 
-    if(m_isVerbose)
-    {
-      dumpNode(xfer_node, fmt::format("round_{}_r{}_begin.json", 0, m_rank));
+      if(m_isVerbose)
+      {
+        dumpNode(xferNode, fmt::format("round_{}_r{}_begin.json", 0, m_rank));
+      }
     }
 
-    // Find initial values on this rank
-    switch(m_runtimePolicy)
-    {
-    case RuntimePolicy::seq:
-      computeLocalClosestPoints<SeqBVHTree>(m_bvh_seq.get(), xfer_node, true);
-      break;
+    /*
+      Send query partition to other processes for searches.  When it
+      returns, search against local object partition.  Query partitions
+      are sent in a ring pattern.  If an object partition is farther
+      than the thresshold distance from the query partition, skip over
```

Comment:
Rather than a ring algorithm, I had originally envisioned each rank having a list (array) of ranks that it needed to send information to, and a list (array) of ranks that it needed to receive information from.

These lists could be generated in a function from the `allgather`'d bounding boxes and the distance threshold and synchronized across the ranks.

Do you think implementing this function in terms of those lists would lead to a cleaner function?

I'm wondering specifically in terms of (a) maintenance and (b) preparedness for the future where will need to support "domain overloading" (where each ranks can have multiple "object" meshes and associated BVH trees / bounding boxes)

Reply:
Yes, I think that would lead to a cleaner function.  For now, I can re-organize the code a bit to improve readability.  The current algorithm does have the potential to reduce communication volume further, by comparing bounding box distances to the closest distances found so far.  I'm not sure that the gain is worth the complexity though.

I would like to better understand how domain overloading can affect either algorithm.

Reply:
Sure -- that can wait for the next PR.